microsoft/testsuites/core/storage.py

```diff
@@ -242,6 +245,155 @@ def verify_os_partition_identifier(self, log: Logger, node: RemoteNode) -> None:
                 "'PARTUUID={os_partition_info.part_uuid} /' should be present in fstab"
             )
 
+    @TestCaseMetadata(
+        description="""
+        This test case will verify that the disks can be added serially while the
+        vm is running.
+        Steps:
+        1. Get maximum number of data disk for the current vm_size.
+        2. Get the number of data disks already added to the vm.
+        3. Serially add and remove the data disks and verify that the added disks
+        are present in the vm.
+        """,
+        requirement=simple_requirement(supported_platform_type=[AZURE]),
+    )
+    def hot_add_disk_serial(
+        self, log: Logger, node: Node, environment: Environment
+    ) -> None:
+        disk_size_in_gb = 20
+        disk = node.features[Disk]
+        assert isinstance(disk, AzureDisk)
+
+        # get max data disk count for the node
+        max_data_disk_count = disk.get_max_data_disk_count()
+        log.debug(f"max_data_disk_count: {max_data_disk_count}")
+
+        # get the number of data disks already added to the vm
+        assert node.capability.disk
+        assert isinstance(node.capability.disk.data_disk_count, int)
+        current_data_disk_count = node.capability.disk.data_disk_count
+        log.debug(f"current_data_disk_count: {current_data_disk_count}")
+
+        # disks to be added to the vm
+        disks_to_add = max_data_disk_count - current_data_disk_count
+
+        # get partition info before adding data disk
+        partitions_before_adding_disk = node.tools[Lsblk].get_partitions(force_run=True)
+
+        for i in range(disks_to_add):
+            disk_id = self._get_managed_disk_id(str(i))
+
+            # add data disk
+            log.debug(f"Adding managed disk: {disk_id}")
+            disk.add_managed_data_disk(
+                [disk_id], [str(i + current_data_disk_count)], size=str(disk_size_in_gb)
+            )
+
+            # verify that partition count is increased by 1
+            # and the size of partition is correct
+            partitons_after_adding_disk = node.tools[Lsblk].get_partitions(
+                force_run=True
+            )
+            added_partitions = [
+                item
+                for item in partitons_after_adding_disk
+                if item not in partitions_before_adding_disk
+            ]
+            log.debug(f"added_partitions: {added_partitions}")
+            assert_that(added_partitions).is_length(1)
+            assert_that(added_partitions[0].size).is_equal_to(
+                disk_size_in_gb * 1024 * 1024 * 1024
+            )
+
+            # remove data disk
+            log.debug(f"Removing managed disk: {disk_id}")
+            disk.remove_managed_data_disk([disk_id])
```

Comment:
It should be called in `after_case`. So if the test failed on previous steps, they can be removed too.

Reply:
This is part of the test case itself. The test basically check that disks can be removed. How about I send `BadEnvironmentStateException` if this steps fails for any reason?

Reply:
The `after_case` will be called after each test case run, it's designed for the cleanup purpose. raise exception overwrites the original failure exception and it needs a big try...except....

Reply:
I need to run `disk.remove_managed_data_disk` as part of the test. There is a check after this which calls `lsblk` to verify that disks have been successfully removed. From a naming point of view, `after_case` sounds something where we just put cleanup operations that are not directly related to the test. Maybe we can discuss this offline to reach a get more clarity